TilerServer/TilerWebModule.pas

```diff
@@ -3698,69 +3703,518 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
+
+                      polyPoints := [xPrev, yPrev, xCurr, yCurr, xCurr-xd, yCurr-yd, xPrev-xd, yPrev-yd];
+                      polygon := ConstructPolygon(polyPoints);
+
+                      DrawFillPolygon(colorsR, aBitmap, polygon);
+                    end;
+                  end;
+                end;
+              end;
+            end;
+          end;
+        end;
+        // todo: process ref geometries not in current? or just skip..
+      finally
+        fCurrentSlice.fDataLock.EndRead;
+        fRefSlice.fDataLock.EndRead;
+      end;
+      Result := gtsOk;
+    finally
+      aBitmap.Canvas.EndScene;
+    end;
+  end
+  else Log.WriteLn('TSliceDiffGeometryICLR layer '+fLayer.LayerID.ToString+': no palette defined', llError);
+end;
+
+function TSliceDiffGeometryICLR.CalculateWidth(aActiveValue, aRefValue: Double; var aValidFlag: Boolean): Double;
+begin
+  if not IsNaN(aActiveValue) then
+  begin
+    if not IsNaN(aRefValue) then
+      Result := (aRefValue + aActiveValue) / 2
+    else Result := aActiveValue;
+  end
+  else
+  begin
+    if not IsNaN(aRefValue)
+    then Result := aRefValue
+    else aValidFlag := False;
+  end;
+end;
+
+function TSliceDiffGeometryICLR.GetPaletteColor(aActiveTexture, aRefTexture: Double; var aValidFlag: Boolean): TGeoColors;
+begin
+  if not (IsNaN(aActiveTexture) or IsNaN(aRefTexture))then
+    Result := fPalette.ValueToColors(aActiveTexture-aRefTexture)
+  else aValidFlag := False;
+end;
+
+function TSliceDiffGeometryICLR.ConstructPolygon(aPolyPoints: TArray<Double>): TPolygon;
+var
+  polygon: TPolygon;
+  ele: Integer;
+begin
+  setLength(polygon, 5);
+  ele := 0;
+  While ele < length(aPolyPoints) do
+  begin
+    polygon[trunc(ele/2)].X := aPolyPoints[ele];
+    inc(ele,1);
+    polygon[trunc(ele/2)].Y := aPolyPoints[ele];
+    inc(ele,1);
+  end;
+
+  polygon[trunc(ele/2)].X := aPolyPoints[0];
+  ele := ele+1;
+  polygon[trunc(ele/2)].Y := aPolyPoints[1];
+
+  Result := polygon;
+end;
+
+procedure TSliceDiffGeometryICLR.DrawFillPolygon(aColor: TGeoColors; aBitmap: FMX.Graphics.TBitmap; aPolygon: TPolygon);
+begin
+  if aColor.fillColor<>0  then
+  begin
+      aBitmap.Canvas.Fill.Color := aColor.fillColor;
+      aBitmap.Canvas.FillPolygon(aPolygon, 1);
+  end;
+  if aColor.outlineColor<>0 then
+  begin
+      aBitmap.Canvas.Stroke.Color := aColor.outlineColor;
+      aBitmap.Canvas.DrawPolygon(aPolygon, 1);
+  end;
+end;
+
+{ TSliceDiffGeometryICLR2 }
+
+constructor TSliceDiffGeometryICLR2.Create(aLayer: TLayer; aPalette: TWDPalette; aTimeStamp: TDateTime; aCurrentSlice, aRefSlice: TSliceGeometryICLR);
+begin
+  inherited Create(aLayer, aPalette, aTimeStamp, aCurrentSlice, aRefSlice);
+end;
+
+destructor TSliceDiffGeometryICLR2.Destroy;
+begin
+  inherited;
+end;
+
+function TSliceDiffGeometryICLR2.GenerateTileCalc(const aExtent: TExtent; aBitmap: FMX.Graphics.TBitmap; aPixelWidth, aPixelHeight: Double): TGenerateTileStatus;

```

Comment:
This method is too big to easily understand it. Please split it on a number of simpler methods. You already put some comments within the method, this is a very good indicator that you have logical pieces that should be moved into separate methods.

Reply:
Done!